packages/node/src/channels.ts

```diff
@@ -407,46 +233,115 @@ export class Channels {
-      if (channel.proposedAppInstances) {
-        apps.push(...Object.values(channel.proposedAppInstances));
-      } else {
-        console.log(
-          `No app instances exist for channel with multisig address: ${
-            channel.multisigAddress
-          }`
-        );
-      }
-    });
-    return apps;
-  }
+  private createAppInstanceFromAppInstanceInfo(
+    proposedAppInstanceInfo: ProposedAppInstanceInfo,
+    channel: StateChannel
+  ): AppInstance {
+    const appInstanceInfo = { ...proposedAppInstanceInfo };
+    const appFunctionSigHashes = this.getAppFunctionSigHashes(appInstanceInfo);
+
+    const appInterface: AppInterface = {
+      addr: appInstanceInfo.appId,
+      applyAction: appFunctionSigHashes.applyAction,
+      resolve: appFunctionSigHashes.resolve,
+      getTurnTaker: appFunctionSigHashes.getTurnTaker,
+      isStateTerminal: appFunctionSigHashes.isStateTerminal,
+      stateEncoding: appInstanceInfo.abiEncodings.stateEncoding,
+      actionEncoding: appInstanceInfo.abiEncodings.actionEncoding
+    };
+
+    const terms: Terms = {
+      assetType: appInstanceInfo.asset.assetType,
+      limit: appInstanceInfo.myDeposit.add(appInstanceInfo.peerDeposit)
+    };
+    if (appInstanceInfo.asset.token) {
+      terms.token = appInstanceInfo.asset.token;
+    } else {
+      terms.token = AddressZero;
+    }
 
-  private async getChannelFromClientAppInstanceID(
-    clientAppInstanceID: string
-  ): Promise<Channel> {
-    const multisigAddress = await this.getMultisigAddressFromClientAppInstanceID(
-      clientAppInstanceID
-    );
-    const channel = await this.getChannelJSONFromStore(multisigAddress);
-    return new Channel(
+    return new AppInstance(
       channel.multisigAddress,
+      // TODO: generate ephemeral app-specific keys
       channel.multisigOwners,
-      channel.rootNonce,
-      channel.freeBalances,
-      channel.appInstances,
-      channel.proposedAppInstances
+      appInstanceInfo.timeout.toNumber(),
+      appInterface,
+      terms,
+      // TODO: pass correct value when virtual app support gets added
+      false,
+      // TODO: this should be thread-safe
+      channel.numInstalledApps,
+      channel.rootNonceValue,
+      appInstanceInfo.initialState,
+      0,
+      appInstanceInfo.timeout.toNumber()
     );
   }
 
-  private generateNewMultisigAddress(owners: Address[]): Address {
-    // TODO: implement this using CREATE2
-    return Wallet.createRandom().address;
+  async getAppInstanceInfoFromAppInstance(
+    appInstances: AppInstance[]
+  ): Promise<AppInstanceInfo[]> {
+    const appInstanceInfos: AppInstanceInfo[] = [];
+    for (const appInstance of appInstances) {
+      const appInstanceId = await this.store.getAppInstanceIDFromAppInstanceIdentityHash(
+        appInstance.identityHash
+      );
+      appInstanceInfos.push(await this.store.getAppInstanceInfo(appInstanceId));
+    }
+    return appInstanceInfos;
```

Comment:
I think this function could be written more clearly using `Array.prototype.map` on `appInstances`.